features/config/service/impl/src/main/java/org/opennms/features/config/service/impl/ConfigurationManagerServiceImpl.java

```diff
@@ -28,136 +28,69 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.stereotype.Component;
 
+import javax.xml.bind.JAXBException;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
 @Component
 public class ConfigurationManagerServiceImpl implements ConfigurationManagerService {
     private final static Logger LOG = LoggerFactory.getLogger(ConfigurationManagerServiceImpl.class);
     private final ConfigStoreDao<JSONObject> configStoreDao;
     // This map contains key: ConfigUpdateInfo value: list of Consumer
     private final ConcurrentHashMap<ConfigUpdateInfo, Collection<Consumer<ConfigUpdateInfo>>> onloadNotifyMap = new ConcurrentHashMap<>();
 
-    private final Map<String, ConfigDefinition> configDefinitions = new HashMap<>(); // TODO: Patrick: need to be replaced with proper implementations
-
     public ConfigurationManagerServiceImpl(final ConfigStoreDao<JSONObject> configStoreDao) {
         this.configStoreDao = configStoreDao;
     }
 
     @Override
-    public void registerSchema(String configName, String xsdName, String topLevelElement)
-            throws IOException, JAXBException {
-        XmlConverter converter = new XmlConverter(xsdName, topLevelElement);
-        Objects.requireNonNull(configName);
-        Objects.requireNonNull(converter);
-        if (this.getRegisteredSchema(configName).isPresent()) {
-            throw new IllegalArgumentException(String.format("Schema with id=%s is already registered.", configName));
-        }
-        final ConfigSchema configSchema = new ConfigSchema(configName, converter.getClass(), converter);
-        configStoreDao.register(configSchema);
-    }
-
-    @Override
-    public void registerConfigDefinition(String configName, ConfigDefinition configDefinition) {
-        // TODO: Patrick: implement me properly
+    public void registerConfigDefinition(String configName, ConfigDefinition configDefinition) throws JsonProcessingException {
         Objects.requireNonNull(configName);
         Objects.requireNonNull(configDefinition);
-        if (this.getRegisteredSchema(configName).isPresent() || configDefinitions.containsKey(configName)) {
+
+        if (this.getRegisteredConfigDefinition(configName).isPresent()) {
             throw new IllegalArgumentException(String.format("Schema with id=%s is already registered.", configName));
         }
-        this.configDefinitions.put(configName, configDefinition); // TODO: Patrick fix this mock:
         try {
-            ConfigSchema<?> schema = new ConfigSchema<>(configDefinition.getConfigName(),
-                    XmlConverter.class,
-                    new XmlConverter("provisiond-configuration.xsd", "provisiond-configuration"));
-            configStoreDao.register(schema);
-        } catch(IOException | JAXBException e) {
+            configStoreDao.register(configDefinition);
```

Comment:
I think we are missing the validation:

Each existing config should be validated against the new schema and only if that works the schema should be imported.

Reply:
Do I need to provide an override option for you? Let say it is a major config change, schema and config will both update together.

Reply:
Anyway added a checking, may be when problem exist add a new logic to handle.

Reply:
Hm this is a good question. Maybe we need to put the validation at the end of the changeset. But: For now I would post phone that until we run into that problem and just validate the existing data against the new schema